### ISSUE TITLE
`Dim` no longer has `dimension` attribute

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#1265]
Code in in the convolution layer in `returnn_common` is trying to set the `dimension` attribute of a `Dim` tag, which no longer exists after the last re-work. I assume we are missing a setter here?

```python
  File "/u/rossenbach/experiments/tts_asr_2021/recipe/returnn_common/nn/conv.py", line 479, in make_conv_out_spatial_dims
    line: out_spatial_dims[i].dimension = ConvLayer.calc_out_dim(
            in_dim=in_spatial_dims[i].dimension,
            filter_size=filter_size[i],
            stride=strides[i],
            dilation_rate=dilation_rate[i],
            padding=padding)
    locals:
      out_spatial_dims = <local> [Dim{'conformer_aed_model/encoder/input_layer/pool1d/_pool_nd:out-spatial-dim0'[?]}]
      i = <local> 0
      dimension = <not found>
      ConvLayer = <local> <class 'returnn.tf.layers.basic.ConvLayer'>
      ConvLayer.calc_out_dim = <local> <bound method ConvLayer.calc_out_dim of <class 'returnn.tf.layers.basic.ConvLayer'>>
      in_dim = <not found>
      in_spatial_dims = <local> [Dim{F'extractor_mel_feature'(80)}]
      filter_size = <local> [2]
      stride = <not found>
      strides = <local> [2]
      dilation_rate = <local> [1]
      padding = <local> 'same'
AttributeError: can't set attribute

```

**Comment by JackTemaki** [comment]
I would have tried to fix it myself, but it is not clear how this should be resolved. I understand that `dimension` is something that should no longer be used explicitly, and this information is now distributed among `capacity`, `size` and `dyn_size_ext` (this variable should definitely be renamed).

I now see two options:
 1. There should be a setter which has the logic from `dimension` to the three variables just as the current constructor has.
 2. `dimension` should not used anymore and the code in the convolution layer should be changed accordingly.
 
Unfortunately it is not clear from the code which of these two options (or even a third one I do not see) is the preferable one to solve this issue.

**Comment by albertz** [comment]
Well, I don't know.

1. First of, the motivation is that we might have static shapes (`capacity`) but with dynamic sizes (`dyn_size_ext`). This is e.g. the case for TPUs but also for ONNX and maybe in other cases. Then `dyn_size_ext` is defined, and `capacity` is set to an integer. `dimension` is None in this case, to indicate that we have dynamic sizes.

2. We might also have the case that some hardware only supports static dimension of power of two for shapes (I think this was the case for the very early TPUs), but the actual dimension you are handling is static but sth else. In this case, `capacity` is this power of two value (basically always what `tensor.shape.as_list()[axis]` would give you), `dimension` is the actual static dimension you wanted to have, `dyn_size_ext` is None (because it is not dynamic).

The first case is not so uncommon for us, and important to support. So far, we did not really support it, and this was the first step towards supporting it. The second case is maybe not really needed. I think current TPUs don't have this limitation anymore. I'm not sure if there is any other case where we might have such limitation.

However, almost all code in RETURNN does not really make those distinctions. Before this, we either had `dimension` set to some integer, meaning that it is static and not dynamic (`dyn_size_ext` not set). Or when `dimension` is None, it mean it is dynamic, and `dyn_size_ext` is supposed to be set. Actually, there is also the state where some dimension is yet undefined, e.g. when you just have `SpatialDim("time")` in the config, and then it would get defined once RETURNN starts using it. Some dimension is undefined when both `dimension` and `dyn_size_ext` are None.

Now with this new distinction between `capacity` and `size`, it's not so clear always in the code which one to use actually. Sometimes you want to get really the underlying shape dim (`tensor.shape.as_list()[axis]` or `tf.shape(tensor)[axis]`), sometimes you want the actual size information (`dyn_size_ext` or `size`).

In your specific example, the code only makes sense if `capacity == size` always, i.e. when the dim is static, the capacity is the same as the size. If this is not true for some static dim, we would not really know what the resulting capacity is. Maybe it is the next higher power of two, but maybe it's also sth else.

Actually, also in the case when the dim is dynamic but capacity is static (TPUs, ONNX), I would not really know how to handle this. How do we know what the resulting tensor shape is? Do we know? We must know this to be able to calculate the output capacity. Or is it just `calc_out_dim(capacity, ...)` in this case? Yes, probably.

Maybe we should just handle `capacity` and `size` separately here. Just as we handled `dimension` and `dyn_size_ext` separately before.

To support legacy code, and the much more common case we usually have, it might make sense to introduce a `dimension` setter though, which behaves the same way as when you pass `dimension` to the constructor. So far, while do the refactoring in RETURNN, I just always set both `capacity` and `size` to the same value.

**Comment by albertz** [comment]
Btw, none of the tests in RETURNN-common are broken. It seems your case is not covered by the tests? It would be good if you can make a simple test case.

**Comment by JackTemaki** [comment]
Yes, I will take of this and other issues after the submission deadline, beginning of next week the latest.
